### PR TITLE
Launch Checklist Issue Template

### PR DESCRIPTION
**Pull request by adamzimmermann**
## Description
Adds a launch checklist issue type.

## Motivation / Context
We need a standardized set of things to check as we prepare to launch sites.

## Testing Instructions / How This Has Been Tested
N/A

## Screenshots
N/A

## Documentation
This is the documentation.

## PR Author Checklist
- [ ] ~~My PR description includes a reference to a ticket or other "motivation" for these changes.~~
- [ ] ~~My PR description includes details of how I tested these changes.~~
- [ ] ~~My PR description includes testing instructions for reviewers with links to the testing environment, if applicable.~~
- [ ] ~~My code includes documentation updates to existing documentation or new sections if applicable.~~

### CODE REVIEW

**Comment by adamzimmermann** [comment]
@JenelleFord FYI this is in progress. More updates to come later today and tomorrow.

**Comment by adamzimmermann** [comment]
@agarzola @markdorison this is a lot better than what we had, but I think it could still be improved.

So I want to keep working on this, but wanted to get your eyes on this to help me spot areas I have haven't even thought of.

**Comment by adamzimmermann** [comment]
@JenelleFord if this is needed now, it could be used as is. However, if it could go through some review next week, I think we could make it even better.

**Comment by mattdaniel** [comment]
Great work @adamzimmermann . One thing that I was thinking we should add is a "Lower DNS TTL" task, perhaps in pre-launch to ensure that the "Change DNS" task propagates quickly... thoughts?

**Comment by JenelleFord** [comment]
@adamzimmermann it's not needed yet, let's take it through the team review before implementing it. Great work on this, thank you!

**Comment by adamzimmermann** [comment]
@mattdaniel 💯 

I'll get that added.

Thank you!

File: .github/ISSUE_TEMPLATE/launch.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+## Preparation
+- [ ] Determine production domain.
+- [ ] Determine if any existing traffic needs to be proxied or redirected.
+- [ ] Determine how redirects will be handled if they are needed.
+- [ ] Determine where the database will come from.
+- [ ] Determine who manages DNS.
+- [ ] Determine the caching strategy.
+
+## Security
+- [ ] Review the Security Review module (https://www.drupal.org/project/security_review).
```
Comment (markdorison) [inline .github/ISSUE_TEMPLATE/launch.md]:
```suggestion
- [ ] Review the [Security Review module](https://www.drupal.org/project/security_review).
```

File: .github/ISSUE_TEMPLATE/launch.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+name: Launch
+about: Prepare for a site launch with a checklist of common steps.
+title: ''
+labels: 'devops'
+assignees: ''
+
+---
+
+Many of the following items can simply be confirmed and checked off. Some may
+warrant their own ticket that references back to a line item here.
```
Comment (markdorison) [inline .github/ISSUE_TEMPLATE/launch.md]:
```suggestion
Please check off line-items as they are completed and leave notes if necessary.
If an item is not relevant to the site launch, [strike it out](https://docs.github.com/en/github/writing-on-github/basic-writing-and-formatting-syntax#styling-text)
ex: `~~Not relevant item~~`. If child tickets are created for any line-item,
please update this ticket description to include references to those tickets.
```

File: .github/ISSUE_TEMPLATE/launch.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+- [ ] Verify that development related accounts/passwords have been reset or removed.
+- [ ] Verify that proper permissions are set on settings and services.yml files.
+- [ ] Confirm that development modules are disabled on produciton
+      (`kint`, `devel_generate`, `devel`, etc).
+- [ ] Verify that `$settings['trusted_host_patterns']` is configured.
+
+## Analytics
+- [ ] Configure Google Tag Manager if applicable.
+- [ ] Configure Google Analytics or your analytics tool of choice.
+- [ ] Configue New Relic if applicable.
```
Comment (markdorison) [inline .github/ISSUE_TEMPLATE/launch.md]:
```suggestion
- [ ] Configure New Relic if applicable.
- [ ] Verify analytics tools are only reporting from the production environment.
```

File: .github/ISSUE_TEMPLATE/launch.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+- [ ] Enable Stage File Proxy for non-production environments.
+- [ ] Verify that config exports and imports cleanly.
+- [ ] Configure external cron trigger (Jenkins/Crontab/etc).
+- [ ] Verify that the deployment process works.
+- [ ] Check site error logs and fix persistent issues.
+- [ ] Verify that Drupal's status report page is clear of issues.
+- [ ] Verify that migrations are complete.
+
+## Launch Prep
+- [ ] Determine who will make the go/no-go call.
```
Comment (markdorison) [inline .github/ISSUE_TEMPLATE/launch.md]:
```suggestion
```

Duplicate. See below.

File: .github/ISSUE_TEMPLATE/launch.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+- [ ] Determine who manages DNS and will do the cutover.
+- [ ] Lower DNS TTL.
+- [ ] Verify that `robots.txt` is configured to allow traffic to the production
+      site (i.e. remove any changes that were made for development).
+- [ ] Verify that users of the site are aware of the cutover time/plan.
+
+## Site Launch
+- [ ] Disable http-auth if applicable.
+- [ ] Change DNS.
+= [ ] Smoke test the site for egregious issues.
```
Comment (markdorison) [inline .github/ISSUE_TEMPLATE/launch.md]:
```suggestion
- [ ] Test the site for egregious issues.
```

File: .github/ISSUE_TEMPLATE/launch.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+- [ ] Determine who will determine if a roll-back is needed.
+- [ ] Determine who manages DNS and will do the cutover.
+- [ ] Lower DNS TTL.
+- [ ] Verify that `robots.txt` is configured to allow traffic to the production
+      site (i.e. remove any changes that were made for development).
+- [ ] Verify that users of the site are aware of the cutover time/plan.
+
+## Site Launch
+- [ ] Disable http-auth if applicable.
+- [ ] Change DNS.
```
Comment (markdorison) [inline .github/ISSUE_TEMPLATE/launch.md]:
```suggestion
- [ ] Update DNS entries.
```

File: .github/ISSUE_TEMPLATE/launch.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+- [ ] Verify that migrations are complete.
+
+## Launch Prep
+- [ ] Determine who will make the go/no-go call.
+- [ ] Create a roll-back plan.
+- [ ] Notify the devops team and confirm they understands what they need to do.
+- [ ] Create ordered checklist and timeline.
+- [ ] Determine who will make the go/no-go call.
+- [ ] Determine who will determine if a roll-back is needed.
+- [ ] Determine who manages DNS and will do the cutover.
```
Comment (markdorison) [inline .github/ISSUE_TEMPLATE/launch.md]:
```suggestion
- [ ] Determine who manages DNS and will update the entries for launch.
```

File: .github/ISSUE_TEMPLATE/launch.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+- [ ] Check site error logs and fix persistent issues.
+- [ ] Verify that Drupal's status report page is clear of issues.
+- [ ] Verify that migrations are complete.
+
+## Launch Prep
+- [ ] Determine who will make the go/no-go call.
+- [ ] Create a roll-back plan.
+- [ ] Notify the devops team and confirm they understands what they need to do.
+- [ ] Create ordered checklist and timeline.
+- [ ] Determine who will make the go/no-go call.
```
Comment (markdorison) [inline .github/ISSUE_TEMPLATE/launch.md]:
```suggestion
- [ ] Determine who will make the go/no-go call for launch.
```

**Review by markdorison — CHANGES_REQUESTED** [review]

File: .github/ISSUE_TEMPLATE/launch.md (hunk 2 above)
Comment (adamzimmermann) [reply .github/ISSUE_TEMPLATE/launch.md]: 🔥

**Review by adamzimmermann — COMMENTED** [review]
@markdorison thanks for the great feedback.

**Comment by adamzimmermann** [comment]
I have a lot here, but I'm sure I'm missing stuff. Please find the areas I missed, as we all bring unique experience with lessons learned from previous site launches.

**Comment by adamzimmermann** [comment]
@JenelleFord are there any special PM related concerns that we should address here too?

File: .github/ISSUE_TEMPLATE/launch.md [hunk 9]
```diff
@@ -0,0 +1,154 @@
+---
+name: Launch
+about: Prepare for a site launch with a checklist of common steps.
+title: ''
+labels: 'devops'
```
Comment (agarzola) [inline .github/ISSUE_TEMPLATE/launch.md]: Should we add the `dept: devops 🚧` label as well? My guess is that repos that don’t have a label called `devops` will simply not apply the label. Same goes for `dept: devops 🚧`.

Thoughts?

File: .github/ISSUE_TEMPLATE/launch.md [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+## Performance
+- [ ] Enable Drupal Page caching.
+- [ ] Enable Drupal Big Pipe caching if needed.
+- [ ] Verify that custom render array cache metadata is set and tested.
+- [ ] Determine if CSS/JavaScript aggregation should be enabled.
+- [ ] Verify that development related services are disabled (e.g.
+      `development.services.yml`)
+- [ ] Verify Redis connectivity and Redis is used for caching.
+- [ ] Verify that Redis is using the correct cache prefix if applicable.
+- [ ] Configured and test the CDN if applicable.
```
Comment (agarzola) [inline .github/ISSUE_TEMPLATE/launch.md]:
```suggestion
- [ ] Configure and test the CDN if applicable.
```

File: .github/ISSUE_TEMPLATE/launch.md [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+- [ ] Verify the CDN is serving HIT's if applicable.
+- [ ] Verify Varnish is serving HIT's if applicable.
```
Comment (agarzola) [inline .github/ISSUE_TEMPLATE/launch.md]: Pluralized acronyms don’t require an apostrophe. I suggest removing them:
```suggestion
- [ ] Verify the CDN is serving HITs if applicable.
- [ ] Verify Varnish is serving HITs if applicable.
```

File: .github/ISSUE_TEMPLATE/launch.md [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+- [ ] Verify that users of the site are aware of the cutover time/plan.
+- [ ] Create a calendar invite with a video call link that can be shared with
+      everyone who is involved with the launch.
+- [ ] Verify that user accounts have been migrated or recreated, and that users
+      have access they need.
+
+## Site Launch
+- [ ] Disable HTTP Authorization if applicable.
+- [ ] Update DNS entries.
+- [ ] Test the site for egregious issues.
```
Comment (agarzola) [inline .github/ISSUE_TEMPLATE/launch.md]: This feels vague. Do you mean performing thorough manual testing of the site in a browser to make sure nothing is obviously broken? If so, I might suggest rewording this to be more explicit.

File: .github/ISSUE_TEMPLATE/launch.md [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+
+## Post Launch
+- [ ] Monitor site log for errors.
```
Comment (agarzola) [inline .github/ISSUE_TEMPLATE/launch.md:147]: Which logs are we talking about here?

File: .github/ISSUE_TEMPLATE/launch.md [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+- [ ] Verify caching is working as intended.
+- [ ] Verify analytics is reporting as expected.
+- [ ] Run performance tests to confirm the site works as expected.
```
Comment (agarzola) [inline .github/ISSUE_TEMPLATE/launch.md]:
```suggestion
- [ ] Run performance tests to confirm the site performs as expected.
```

Is this referring to running a test with a tool such as Calibre, Lighthouse, or PageSpeed Insights?

File: .github/ISSUE_TEMPLATE/launch.md [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+- [ ] Update QA environments to pull databases from the new production site.
+- [ ] Update Stage File Proxy origin URL configuration to point to the
+      production domain (e.g. `$config['stage_file_proxy.settings']['origin']`).
+- [ ] Update automated or manual tests to point to the new production domain.
+- [ ] Confirm that third party integrations are working as expected.
+- [ ] Disable migration related modules and database connectivity as applicable.
```
Comment (agarzola) [inline .github/ISSUE_TEMPLATE/launch.md]: I take it this refers to source database connectivity? If so, I suggest explicitly calling it that:
```suggestion
- [ ] Disable migration related modules and source database connectivity as applicable.
```

**Review by agarzola — COMMENTED** [review]
This is looking great! See comments/suggestions/questions below.

File: .github/ISSUE_TEMPLATE/launch.md (hunk 9 above)
Comment (adamzimmermann) [reply .github/ISSUE_TEMPLATE/launch.md]: I've been trying to avoid adding labels to create dependencies (even though I don't think it hurts anything), as it's not that hard to apply your own labels, especially when you only create this issue once per site. So I'm inclined to say no, but also 🤷‍♂️ .

File: .github/ISSUE_TEMPLATE/launch.md (hunk 14 above)
Comment (adamzimmermann) [reply .github/ISSUE_TEMPLATE/launch.md]: Yes. Adding a reference to them.

File: .github/ISSUE_TEMPLATE/launch.md (hunk 13 above)
Comment (adamzimmermann) [reply .github/ISSUE_TEMPLATE/launch.md:147]: Whatever ones we deem important? I could say PHP logs, but I kind of like this being vague. Thoughts?

File: .github/ISSUE_TEMPLATE/launch.md (hunk 12 above)
Comment (adamzimmermann) [reply .github/ISSUE_TEMPLATE/launch.md]: Yes. Updating.

File: .github/ISSUE_TEMPLATE/launch.md [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+      launch.
+- [ ] Verify that no API keys or sensitive credentials are stored in code.
+- [ ] Verify if "helper" node types and/or taxonomy term pages should appear and
+      hide or block access as needed.
+- [ ] Verify that alias patterns and updates/redirects are working correctly.
+- [ ] Verify that `$settings['hash_salt']` is set.
+- [ ] Verify that `$settings['update_free_access'] = FALSE;` is set.
+- [ ] Verify that `$settings['omit_vary_cookie']` is configured correctly.
+- [ ] Verify that `$settings['file_public_path']` and `file_private_path` are
+      configured correctly.
```
Comment (mmatsoo) [inline .github/ISSUE_TEMPLATE/launch.md:92]: Perhaps we should include a twig method whitelist from the get-go?

```php
$settings['twig_sandbox_whitelisted_methods'] = [
  'id',
  'label',
  'bundle',
  'get',
  '__toString',
  'toString',
  'referencedEntities',
];
```

File: .github/ISSUE_TEMPLATE/launch.md (hunk 13 above)
Comment (agarzola) [reply .github/ISSUE_TEMPLATE/launch.md:147]: I worry that it’s vague enough to gloss over, but I don’t feel too strongly about this. I’m cool with leaving it as is.

File: .github/ISSUE_TEMPLATE/launch.md (hunk 9 above)
Comment (agarzola) [reply .github/ISSUE_TEMPLATE/launch.md]: I get that. `dept: devops 🚧` seems to be a standard we use across several projects (do we import those from somewhere?), so I think it makes about as much sense as `devops`. That said, I don’t feel too strongly about this. I’m fine with leaving this as is.

**Review by agarzola — APPROVED** [review]
Looks good to me!

File: .github/ISSUE_TEMPLATE/launch.md (hunk 9 above)
Comment (markdorison) [reply .github/ISSUE_TEMPLATE/launch.md]: > I get that. dept: devops 🚧 seems to be a standard we use across several projects (do we import those from somewhere?)

The problem as I see it is that labels are not standardized across projects. They sometimes vary wildly but more often just in small ways (emoji or no emoji). There was not previously a way to standardize labels, but in my due diligence while writing this comment, I discovered that [GitHub rolled out some functionality around this last year: [Managing default labels for repositories in your organization](https://docs.github.com/en/github/setting-up-and-managing-organizations-and-teams/managing-default-labels-for-repositories-in-your-organization). This will only work for new repos though, we would have to manually update existing ones.

These are our options as I see it.

1. Include no labels in the template. Let repos specify their own.
1. Include a sane default. If there is not a matching label in the repo, it will just ignore it. In those cases, it is effectively the same as option 1.
1. Include multiple derivatives of each label. If any match they will apply those.
1. [Standardize our labels](https://docs.github.com/en/github/setting-up-and-managing-organizations-and-teams/managing-default-labels-for-repositories-in-your-organization) for new repos and manually update existing ones and then come back and update this template once that is done.

File: .github/ISSUE_TEMPLATE/launch.md [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+- [ ] Run load tests to validate infrastructure.
+- [ ] Consult with the hosting provider for any platform specific launch steps
+      or advice.
+
+## Launch Prep
+- [ ] Create a roll-back plan.
+- [ ] Notify the devops team and confirm they understands what they need to do.
+- [ ] Create ordered checklist and timeline.
+- [ ] Determine who will make the go/no-go call for launch.
+- [ ] Determine who will determine if a roll-back is needed.
```
Comment (agarzola) [inline .github/ISSUE_TEMPLATE/launch.md]: I just noticed this double `determine`.
```suggestion
- [ ] Determine who will decide if a roll-back is needed.
```

File: .github/ISSUE_TEMPLATE/launch.md (hunk 9 above)
Comment (adamzimmermann) [reply .github/ISSUE_TEMPLATE/launch.md]: @markdorison beat me to it, but I was going to propose option 1. I didn't realize that I had applied the `devops` label.

I really don't care too much though, and option 1 seems like the path of least resistance.

Comment (markdorison) [reply .github/ISSUE_TEMPLATE/launch.md]: That sounds good to me.

Comment (agarzola) [reply .github/ISSUE_TEMPLATE/launch.md]: I like option no. 4 the most, but that’s the path of most resistance and can be handled at a later date. For now, I think option no. 3 is my second favorite.

But again, if we leave it as-is I’m cool with it.

File: .github/ISSUE_TEMPLATE/launch.md (hunk 16 above)
Comment (adamzimmermann) [reply .github/ISSUE_TEMPLATE/launch.md:92]: This is more functionality development related vs launch development.

File: .github/ISSUE_TEMPLATE/launch.md (hunk 9 above)
Comment (adamzimmermann) [reply .github/ISSUE_TEMPLATE/launch.md]: I just went with 1 for now. Follow-up PR's are welcome. We can discuss in our upcoming call too.

**Comment by adamzimmermann** [comment]
@markdorison can you give this a ✅ ?

**Review by markdorison — APPROVED** [review]
Sorry for the delay.
